plugins/syntax/numeral.ml

```diff
@@ -128,13 +128,212 @@ let warn_deprecated_decimal =
               Decimal.int or Decimal.decimal. Use Numeral.uint, \
               Numeral.int or Numeral.numeral respectively.")
 
-let vernac_numeral_notation local ty f g scope opts =
+let pr_constr env sigma c =
+  let c = Constrextern.extern_constr env sigma (EConstr.of_constr c) in
+  Ppconstr.pr_constr_expr env sigma c
+
+let remapping_error ?loc env sigma ty ty' ty'' =
+  let constr = pr_constr env sigma in
+  CErrors.user_err ?loc
+    (constr ty ++ str " was already mapped to" ++ spc () ++ constr ty'
+     ++ str " and cannot be remapped to" ++ spc () ++ constr ty'' ++ str ".")
+
+let type_error_using ?loc env sigma c c' t =
+  let constr = pr_constr env sigma in
+  CErrors.user_err ?loc
+    (constr c ++ str " is incompatible with" ++ spc () ++ constr c'
+     ++ str "." ++ spc () ++ str "It's expected type is: " ++ constr t ++ str ".")
+
+let error_missing env sigma c =
+  CErrors.user_err
+    (str "Missing mapping for constructor " ++ pr_constr env sigma c ++ str ".")
+
+let via_abstract_error () =
+  CErrors.user_err (Pp.str "'via' and 'abstract' cannot be used together.")
+
+let locate_global_sort_inductive_or_constant sigma qid =
+  let locate_sort qid =
+    match Nametab.locate_extended qid with
+    | Globnames.TrueGlobal _ -> raise Not_found
+    | Globnames.SynDef kn ->
+    match Syntax_def.search_syntactic_definition kn with
+    | [], Notation_term.NSort r ->
+       let sigma,c = Evd.fresh_sort_in_family sigma (Glob_ops.glob_sort_family r) in
+       sigma,Constr.mkSort c
+    | _ -> raise Not_found in
+  try locate_sort qid
+  with Not_found ->
+    match Smartlocate.global_with_alias qid with
+    | GlobRef.IndRef i -> sigma, Constr.mkInd i
+    | _ -> sigma, Constr.mkConst (Smartlocate.global_constant_with_alias qid)
+
+let locate_global_constructor_inductive_or_constant qid =
+  let g = Smartlocate.global_with_alias qid in
+  match g with
+  | GlobRef.ConstructRef c -> g, Constr.mkConstruct c
+  | GlobRef.IndRef i -> g, Constr.mkInd i
+  | _ -> g, Constr.mkConst (Smartlocate.global_constant_with_alias qid)
+
+let get_type env sigma c =
+  let t = Retyping.get_type_of env sigma (EConstr.of_constr c) in
+  EConstr.to_constr sigma t
+
+let elaborate_to_post_params env sigma ty_ind params =
+  match params with
+  | [] -> [||], get_constructors ty_ind
+  | _ ->
+     let pt_refs = get_constructors ty_ind in
+     let rec prod_length c = match Constr.kind c with
+       | Constr.Prod (_, _, c) -> 1 + prod_length c
+       | _ -> 0 in
+     let l =
+       let args = List.map (fun r -> ToPostCheck r) params in
```

Comment:
This assumes that all constructors have the parameters as their first arguments, right?  So you'll get failures(?) or type errors(?) on things like
```coq
Inductive extra_list : Type -> Type := nil (n : nat) (v : Type) : extra_list v | cons (n : nat) (t : Type) (x : t) : extra_list t -> extra_list t.
```
?